### PR TITLE
[MXNET-1203] Tutorial infogan 

### PR DESCRIPTION
**Pull request by NRauschmayr**
## Description ##

Adding a new tutorial that explains how to do image similarity search based on InfoGAN. https://arxiv.org/abs/1606.03657
@ThomasDelteil @Ishitori @vishaalkapoor can you please have a look? Thanks!

 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)

### CODE REVIEW

File: docs/tutorials/gluon/info_gan.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    visualize(image)
+    for i in range(2,9): 
+        if indices[i-1][1] < 1.5:
+            plt.subplot(1,10,i)
+            sim = test_images[indices[i-1][0]].reshape(3,64,64)
+            visualize(sim)
+    plt.show()
+    plt.clf()
+```
+![png](https://raw.githubusercontent.com/NRauschmayr/web-data/master/mxnet/doc/tutorials/info_gan/output.png)<!--notebook-skip-line-->
```
[the remark this answers is not in the document]
Comment (NRauschmayr) [reply docs/tutorials/gluon/info_gan.md]: Thanks a lot for the review! I already made a pull request to dmlc/web-data. Once it is accepted, I will adjust the links.

[the remark this answers is not in the document]
Comment (NRauschmayr) [reply docs/tutorials/gluon/info_gan.md]: My pull request to dmlc was accepted so I adjusted the links.

File: docs/tutorials/gluon/info_gan.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
+
+# Image similarity search with InfoGAN
+
+This notebook shows how to implement an InfoGAN based on Gluon. InfoGAN is an extension of GANs, where the generator input is split in 2 parts: random noise and a latent code c (see [InfoGAN Paper](https://arxiv.org/pdf/1606.03657.pdf)).
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: and a latent code (remove 'c')?

File: docs/tutorials/gluon/info_gan.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+# Image similarity search with InfoGAN
+
+This notebook shows how to implement an InfoGAN based on Gluon. InfoGAN is an extension of GANs, where the generator input is split in 2 parts: random noise and a latent code c (see [InfoGAN Paper](https://arxiv.org/pdf/1606.03657.pdf)). 
+The codes are made meaningful by maximizing the mutual information between code and generator output. InfoGAN learns a disentangled representation in a completely unsupervised manner. It can be used for many applications such as image similarity search. This notebook uses the DCGAN example from the [Straight Dope Book](https://gluon.mxnet.io/chapter14_generative-adversarial-networks/dcgan.html) and extends it to create an InfoGAN. 
+
+
+```python
+from __future__ import print_function
+from datetime import datetime
+import sys
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: nit, alphabetize

File: docs/tutorials/gluon/info_gan.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
+
+# Image similarity search with InfoGAN
+
+This notebook shows how to implement an InfoGAN based on Gluon. InfoGAN is an extension of GANs, where the generator input is split in 2 parts: random noise and a latent code c (see [InfoGAN Paper](https://arxiv.org/pdf/1606.03657.pdf)). 
+The codes are made meaningful by maximizing the mutual information between code and generator output. InfoGAN learns a disentangled representation in a completely unsupervised manner. It can be used for many applications such as image similarity search. This notebook uses the DCGAN example from the [Straight Dope Book](https://gluon.mxnet.io/chapter14_generative-adversarial-networks/dcgan.html) and extends it to create an InfoGAN. 
+
+
+```python
+from __future__ import print_function
+from datetime import datetime
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: PEP8 doesn't really talk about order of 'from' imports vs 'imports' within the groupings, so I'll leave it to you, but alphabetization is still nicer than no alphabetization :)

File: docs/tutorials/gluon/info_gan.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+from mxnet import gluon
+from mxnet import ndarray as nd
+from mxnet.gluon import nn, utils
+from mxnet import autograd
+from mxboard import SummaryWriter
+import numpy as np
+
+```
+
+The latent code vector c can contain several variables, which can be categorical and/or continuous. We set `n_continuous` to 2 and `n_categories` to 10.
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: is it necessary to mention c when referring generically to the latent code vector? isn't c just a variable used in the paper?

File: docs/tutorials/gluon/info_gan.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+
+```python
+batch_size   = 64
+z_dim        = 100
+n_continuous = 2
+n_categories = 10
+ctx = mx.gpu() if mx.test_utils.list_gpus() else mx.cpu()
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: any reason to use just one of the gpus and not all?

File: docs/tutorials/gluon/info_gan.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+    filenames = []
+    for path, _, fnames in os.walk(data_dir):
+        for fname in fnames:
+            if not fname.endswith('.jpg'):
+                continue
+            img = os.path.join(path, fname)
+            img_arr = mx.image.imread(img)
+            img_arr = transform(img_arr)
+            images.append(img_arr)
+            filenames.append(path + "/" + fname)
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: nit '/'

File: docs/tutorials/gluon/info_gan.md [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+    data = nd.transpose(data, (2,0,1))
+    data = data.astype(np.float32)/127.5 - 1
+    if data.shape[0] == 1:
+        data = nd.tile(data, (3, 1, 1))
+    return data.reshape((1,) + data.shape)
+```
+
+
+```python
+def get_files(data_dir):
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: get_images?

File: docs/tutorials/gluon/info_gan.md [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+            images.append(img_arr)
+            filenames.append(path + "/" + fname)
+    return images, filenames        
+```
+
+Load the dataset `lfw_dataset` which contains images of celebrities.
+
+
+```python
+data_dir          =  'lfw_dataset'
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: nit extra space

File: docs/tutorials/gluon/info_gan.md [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+images, filenames = get_files(data_dir)
+split             = int(len(images)*0.8)
+test_images       = images[split:]
+test_filenames    = filenames[split:]
+train_images      = images[:split]
+train_filenames   = filenames[:split]
+
+train_data        = gluon.data.ArrayDataset(nd.concatenate(train_images))
+train_dataloader  = gluon.data.DataLoader(train_data, batch_size=batch_size, shuffle=True, last_batch='rollover', num_workers=4)
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: the num_workers should be set accordingly, ask @ThomasDelteil :)

File: docs/tutorials/gluon/info_gan.md [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+
+Create vectors with real (=1) and fake labels (=0).
+
+
+```python
+real_label = nd.ones((batch_size,), ctx=ctx)
+fake_label = nd.zeros((batch_size,),ctx=ctx)
+```
+
+Load a pertrained model.
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: pre-trained

File: docs/tutorials/gluon/info_gan.md [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+
+
+```python
+if os.path.isfile("infogan_d_latest.params") and os.path.isfile("infogan_g_latest.params"):
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: nit, consistent use of ' vs "

File: docs/tutorials/gluon/info_gan.md [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+3. Update Discriminator. Currently, it is updated every second iteration in order to avoid that the Discriminator becomes too strong. You may want to change that.
+4. The updated discriminator receives `fake_image` and `loss1` measures how many fake images have been been identified as real, `loss2` measures the difference between the sampled continuous latent code `c` and the output of the Q model and `loss3` measures the difference between the sampled categorical latent code `c` and the output of the Q model.
+4. Update Generator and Q
+
+
+```python
+with SummaryWriter(logdir='./logs/') as sw:
+    
+    epochs = 1
+    i = 0
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: better name for i

also, you don't have to i+=1 and then later check % 2, you can i = (i + 1) % 2 because that's all you care about. it's a nit, but cleaner as i doesn't increase without bound unnecessarily.

File: docs/tutorials/gluon/info_gan.md [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+                
+                # total discriminator error
+                d_error         = d_error_real + d_error_fake
+
+            d_error_epoch += d_error.mean()
+            
+            #Update D every second iteration
+            if i % 2 == 0:
+                d_error.backward()
+                d_trainer.step(data.shape[0])
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: you can define a descriptive variable name for data.shape[0] as it's used several times.

File: docs/tutorials/gluon/info_gan.md [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+                
+                # create some fake image for logging in MXBoard
+                fake_image = generator(g_input)
+
+                sw.add_scalar(tag='Loss_D', value={'test':d_error_epoch.asscalar()/idx}, global_step=i)
+                sw.add_scalar(tag='Loss_G', value={'test':d_error_epoch.asscalar()/idx}, global_step=i)
+                sw.add_image(tag='data_image', image=((fake_image[0]+ 1.0) * 127.5).astype(np.uint8)  , global_step=i)
+                sw.flush()
+                
+            time1 = time.time()
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/info_gan.md]: is time1 used?

**Comment by vishaalkapoor** [comment]
My comments have been addressed, please wait for others to sign off. As noted offline, num_workers = 4 is probably good.

File: docs/tutorials/gluon/info_gan.md [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+        prob            = self.prob(x)
+        feat            = self.feat(x)
+        category_prob   = self.category_prob(feat)
+        continuous_mean = self.continuous_mean(feat)
+        
+        return prob, category_prob, continuous_mean
+```
+
+The InfoGAN has the following layout.
+<img src="https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/info_gan/InfoGAN.png" style="width:800px;height:250px;">
```
Comment (anirudhacharya) [inline docs/tutorials/gluon/info_gan.md]: I think the architecture diagram should be mirror inverted - 

Latent&Noise . -> . Generator . ->  Fake Data( + Real Data) . -> . Discriminator . -> (Real/Fake probability) Q Neural Network . -> Estimation of c

The flow of the diagram should be from left to right

File: docs/tutorials/gluon/info_gan.md [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+Load a pretrained model.
+
+
+```python
+if os.path.isfile('infogan_d_latest.params') and os.path.isfile('infogan_g_latest.params'):
+    discriminator.load_parameters('infogan_d_latest.params', ctx=ctx, allow_missing=True, ignore_extra=True)
+    generator.load_parameters('infogan_g_latest.params', ctx=ctx, allow_missing=True, ignore_extra=True)
+```
+
+The latent code is part of the Generator input and it contains mutliple variables (continuous, categorical) that can represent different distributions. In order to make sure that the Generator uses the latent code, mutual information is introduced into the GAN loss term. Mutual information measures how much X is known given Y or vice versa. It is defined as:
```
Comment (anirudhacharya) [inline docs/tutorials/gluon/info_gan.md]: might be worth mentioning that how InfoGAN architecture varies from DCGAN is that 
 - the Generator has the extra latent code as input a
 - there is an extra Q NN to estimate `c`

File: docs/tutorials/gluon/info_gan.md [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+    label  = nd.array(np.random.randint(n_categories, size=batch_size)).as_in_context(ctx)
+    c1     = nd.one_hot(label, depth=n_categories).as_in_context(ctx)
+    c2     = nd.random.uniform(-1, 1, shape=(batch_size, n_continuous)).as_in_context(ctx)
+
+    # concatenate random noise with c which will be the input of the generator
+    return nd.concat(z, c1, c2, dim=1), label, c2
+```
+
+Define the training loop. 
+1. The discriminator receives `real_data` and `loss1` measures how many real images have been identified as real
```
Comment (anirudhacharya) [inline docs/tutorials/gluon/info_gan.md]: The fact that Discriminator accepts `real data` along with the `fake data` generated by the Generator is currently not reflected in the architecture diagram. Can it be updated?

Comment (NRauschmayr) [reply docs/tutorials/gluon/info_gan.md]: Thanks for the review! I added your comments. I will also update the diagram as you suggested and upload it to dmlc/webdata in the the coming days.

File: docs/tutorials/gluon/info_gan.md (hunk 16 above)
Comment (NRauschmayr) [reply docs/tutorials/gluon/info_gan.md]: I updated the diagram: https://github.com/NRauschmayr/web-data/blob/tutorial_infogan/mxnet/doc/tutorials/info_gan/InfoGAN.png Let me know if you think it should be further improved.

Comment (anirudhacharya) [reply docs/tutorials/gluon/info_gan.md]: Would be better if it showed that the output of the Discriminator is the probability distribution of real and fake, as opposed to it currently showing just "Real|Fake". But its a minor point, you can take a call on what would be better.

**Comment by anirudhacharya** [comment]
LGTM. My comments are addressed. 

@indhub @nswamy please review/merge this PR.

**Comment by NRauschmayr** [comment]
Just following up since the PR is still open. Can we merge or is there anything else I shall update in the tutorial?
@indhub @nswamy @anirudh2290 @szha
